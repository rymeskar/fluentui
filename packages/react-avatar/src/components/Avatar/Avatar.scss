// Define default variable values
:root {
  --avatarSizeSmallest: 20px;
  --avatarSizeSmaller: 24px;
  --avatarSizeSmall: 28px;
  --avatarSizeMedium: 32px;
  --avatarSizeLarge: 44px;
  --avatarSizeLarger: 64px;
  --avatarSizeLargest: 96px;
  --avatarBorderColor: transparent;
  --avatarBorderWidth: 0;
  --avatarBorderRadius: 9999px;

  --squareAvatarBorderRadius: 3px;
  --statusBorderColor: transparent;
  --statusBorderWidth: 2px;
  --avatarFill: rgb(232, 232, 232);
}

// Root slot
.root {
  position: relative;
  background-color: inherit;
  display: inline-block;
  vertical-align: middle;
  flex-shrink: 0;
  width: var(--avatarSize, var(--avatarSizeMedium));
  height: var(--avatarSize, var(--avatarSizeMedium));
}

// Image slot
.image {
  position: absolute;
  border-radius: var(--avatarBorderRadius);
  top: 0;
  left: 0;
  width: var(--avatarSize, var(--avatarSizeMedium));
  height: var(--avatarSize, var(--avatarSizeMedium));
  height: 100%;
  object-fit: cover;
  vertical-align: top;
}

// Label slot
.label {
  display: flex;
  align-items: center;
  justify-content: center;
  background-color: var(--avatarFill);

  font-size: calc(var(--avatarSize) / 2.333);
  width: var(--avatarSize, var(--avatarSizeMedium));
  height: var(--avatarSize, var(--avatarSizeMedium));

  vertical-align: top;
  text-align: center;
  padding: 0px;
  border-radius: 99999px;
}

// Status slot
.status {
  position: absolute;
  bottom: 0;
  right: 0;
  box-shadow: 0 0 0 var(--statusBorderWidth) var(--statusBorderColor);
}

// Root modifier when props.size_ smallest
._size_smallest {
  --avatarSize: var(--avatarSizeSmallest);
}

// Root modifier when props.size_ smaller
._size_smaller {
  --avatarSize: var(--avatarSizeSmaller);
}

// Root modifier when props.size_ small
._size_small {
  --avatarSize: var(--avatarSizeSmall);
}

// Root modifier when props.size_ large
._size_large {
  --avatarSize: var(--avatarSizeLarge);
}

// Root modifier when props.size_ larger
// larger-abc123
._size_larger {
  --avatarSize: var(--avatarSizeLarger);
}

// Root modifier when props.size_ largest
._size_largest {
  --avatarSize: var(--avatarSizeLargest);
}

// Modifiers
._square {
  .image {
    border-radius: var(--squareAvatarBorderRadius);
  }

  .label {
    border-radius: var(--squareAvatarBorderRadius);
  }
}
