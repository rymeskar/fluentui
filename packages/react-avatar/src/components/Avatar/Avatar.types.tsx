import { StatusProps } from '../Status/index';
import { ComponentProps, ShorthandValue } from '../utils/commonTypes';

export type SizeValue = 'smallest' | 'smaller' | 'small' | 'medium' | 'large' | 'larger' | 'largest';

export interface AvatarProps extends ComponentProps, React.HTMLAttributes<HTMLElement> {
  /** Shorthand for the image. */
  image?: ShorthandValue<{}>;

  /** Shorthand for the label. */
  label?: ShorthandValue<{}>;

  /** Shorthand for the status of the user. */
  status?: ShorthandValue<StatusProps>;

  /** The name used for displaying the initials of the avatar if the image is not provided. */
  name?: string;

  /** The avatar can have a square shape. */
  square?: boolean;

  /** Size multiplier. */
  size?: SizeValue;

  /** Custom method for generating the initials from the name property, which is shown if no image is provided. */
  getInitials?: (name: string, isRtl: boolean) => string;

  /** Classes for the parts of the component */
  classes?: { [key: string]: string };
}

export type AvatarState = AvatarProps;
