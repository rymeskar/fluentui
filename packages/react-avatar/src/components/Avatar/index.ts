export * from './Avatar.types';
export * from './useAvatar';
export * from './Avatar';
