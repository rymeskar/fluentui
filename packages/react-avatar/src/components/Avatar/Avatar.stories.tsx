import * as React from 'react';
import { Avatar } from './Avatar';
import { StoryExample } from '../utils/StoryExample';

const imageUrl = 'http://www.fillmurray.com/192/192';

export const AvatarCss = () => (
  <StoryExample title="Avatar (css)">
    <Avatar size="smallest" name="John Doe" status="success" />
    <Avatar size="smaller" name="John Doe" status="success" />
    <Avatar size="small" name="John Doe" status="success" />
    <Avatar name="John Doe" status="success" />
    <Avatar size="large" name="Jane Doe" status="error" />
    <Avatar size="larger" name="Lorem Ipsum" status="warning" />
    <Avatar size="largest" name="Lorem Ipsum" status="warning" />
    <Avatar square size="smallest" name="John Doe" status="success" />
    <Avatar square size="smaller" name="John Doe" status="success" />
    <Avatar square size="small" name="John Doe" status="success" />
    <Avatar square name="John Doe" status="success" />
    <Avatar square size="large" name="Jane Doe" status="error" />
    <Avatar square size="larger" name="Lorem Ipsum" status="warning" />
    <Avatar square size="largest" name="Lorem Ipsum" status="warning" />

    <Avatar size="smallest" name="John Doe" status="success" image={imageUrl} />
    <Avatar size="smaller" name="John Doe" status="success" image={imageUrl} />
    <Avatar size="small" name="John Doe" status="success" image={imageUrl} />
    <Avatar name="John Doe" status="success" image={imageUrl} />
    <Avatar size="large" name="Jane Doe" status="error" image={imageUrl} />
    <Avatar size="larger" name="Lorem Ipsum" status="warning" image={imageUrl} />
    <Avatar size="largest" name="Lorem Ipsum" status="warning" image={imageUrl} />
    <Avatar square size="smallest" name="John Doe" status="success" image={imageUrl} />
    <Avatar square size="smaller" name="John Doe" status="success" image={imageUrl} />
    <Avatar square size="small" name="John Doe" status="success" image={imageUrl} />
    <Avatar square name="John Doe" status="success" image={imageUrl} />
    <Avatar square size="large" name="Jane Doe" status="error" image={imageUrl} />
    <Avatar square size="larger" name="Lorem Ipsum" status="warning" image={imageUrl} />
    <Avatar square size="largest" name="Lorem Ipsum" status="warning" image={imageUrl} />
  </StoryExample>
);
