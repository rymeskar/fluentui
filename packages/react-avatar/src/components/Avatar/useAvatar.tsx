import * as React from 'react';
import { mergeProps, getSlots, resolveShorthandProps } from '@fluentui/react-compose/lib/next/index';
import { AvatarProps, AvatarState } from './Avatar.types';
import { useMergedRefs } from '@uifabric/react-hooks';
import { getInitials, nullRender } from '@uifabric/utilities';
import { Image } from '../Image/index';

const avatarShorthandProps: (keyof AvatarProps)[] = ['label', 'image', 'status'];

export const renderAvatar = (state: AvatarState) => {
  const { slots, slotProps } = getSlots(state, avatarShorthandProps);

  return (
    <slots.root {...slotProps.root}>
      <slots.label {...slotProps.label} />
      <slots.image {...slotProps.image} />
      <slots.status {...slotProps.status} />
    </slots.root>
  );
};

export const useAvatar = (props: AvatarProps, ref: React.Ref<HTMLElement>, defaultProps?: AvatarProps) => {
  const state = mergeProps(
    {
      as: 'span',
      label: {
        as: 'span',
        children: props.getInitials ? props.getInitials(props.name || '', false) : getInitials(props.name || '', false),
      },
      image: { as: props.image ? Image : nullRender },
      status: { as: 'span', size: props.size },
      ref: useMergedRefs(ref, React.useRef(null)),
    },
    defaultProps,
    resolveShorthandProps(props, avatarShorthandProps),
  );

  return { state, render: renderAvatar };
};
