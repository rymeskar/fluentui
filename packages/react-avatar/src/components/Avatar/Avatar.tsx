import * as React from 'react';
import { useAvatar } from './useAvatar';
import { AvatarProps } from './Avatar.types';
import { makeClasses } from '@fluentui/react-compose/lib/next/index';
// import { useInlineTokens } from '@fluentui/react-theme-provider';
import { useFocusRects } from '@uifabric/utilities';
import * as classes from './Avatar.scss';
import { Status } from '../Status/Status';

const useAvatarClasses = makeClasses(classes);

export const Avatar = React.forwardRef((props: AvatarProps, ref: React.Ref<HTMLElement>) => {
  const { state, render } = useAvatar(props, ref, {
    status: { as: Status },
  });

  // Apply styling.
  useAvatarClasses(state);
  useFocusRects(state.ref);

  // AvatarProps.tokens must be defined
  // useInlineTokens(state, '--avatar');

  return render(state);
});

Avatar.displayName = 'Avatar';
