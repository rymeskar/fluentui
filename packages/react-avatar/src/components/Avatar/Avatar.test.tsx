import * as React from 'react';
import * as path from 'path';
import { isConformant } from '@fluentui/react-conformance';
import { Avatar } from './Avatar';
import * as renderer from 'react-test-renderer';
import { ReactWrapper } from 'enzyme';

describe('Avatar', () => {
  let wrapper: ReactWrapper | undefined;

  afterEach(() => {
    if (wrapper) {
      wrapper.unmount();
      wrapper = undefined;
    }
  });

  isConformant({
    componentPath: path.join(__dirname, 'Avatar.tsx'),
    Component: Avatar,
    displayName: 'Avatar',
    disabledTests: ['has-docblock', 'as-renders-html', 'as-passes-as-value', 'as-renders-react-class', 'as-renders-fc'],
  });

  /**
   * Note: see more visual regression tests for Avatar in /apps/vr-tests.
   */
  it('renders a default state', () => {
    const component = renderer.create(<Avatar>Default Avatar</Avatar>);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
