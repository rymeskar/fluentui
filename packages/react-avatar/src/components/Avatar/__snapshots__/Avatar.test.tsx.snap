// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Avatar renders a default state 1`] = `
<span
  className=""
>
  <span>
    
  </span>
  <span
    className=""
  />
</span>
`;
